Abstract interface for disk image

gerph opened the pull request:
# Summary

The disc image being accessed used a FILE * directly. This is fine if
you only want to work with files on disc but if you need to use a
different interface to access the disc image data, this is not much use.

This change introduces a new discimage_t, through which the disc files
are accessed. This means that the access to the disc image can be changed
from being a file to being (say) a block device accessed through a
different interface.

All the internal operations that read or wrote through the standard
FILE * stream interfaces, for the disc image, have now been replaced
with operations that use the discimage_t methods.

# Admin

This change builds upon PR #1.

gerph committed:
Update to be buildable with RISC OS C89 compiler.

The tool now works with the C89 compiler I use, and the limited
getopt that doesn't support long options.

gerph committed:
Add CI build on GitHub + RISC OS Build service.

Add the ability to build through CI on the build service, plus building
on Linux.

gerph committed:
Fix for help text not reflecting short getopt options.

I don't have the long extensions to getopt in the library I'm
using, so I need to make the help reflect that too.

gerph committed:
Add abstract interface for disc image access.

The disc image being accessed used a FILE * directly. This is fine if
you only want to work with files on disc but if you need to use a
different interface to access the disc image data, this is not much use.

This change introduces a new discimage_t, through which the disc files
are accessed. This means that the access to the disc image can be changed
from being a file to being (say) a block device accessed through a
different interface.

All the internal operations that read or wrote through the standard
FILE * stream interfaces, for the disc image, have now been replaced
with operations that use the discimage_t methods.

Cyberspice commented on include/acornfs.h:
```diff
@@ -45,7 +45,7 @@ typedef struct _tag_ACORN_DIRECTORY {
   uint8_t options;
   int num_of_files;
-  ACORN_FILE files[];
+  ACORN_FILE files[1];
```
Doesn't this change the size of ACORN_DIRECTORY? I.e. the structure now includes 1 element of ACORN_FILE, where as before it was only the size of the header? This may not be a problem I'm just trying to understand why the change.

gerph replied:
Oh, yes, it does. And I never corrected for that. I guess that means we return a buffer a little too big.
The reason was that the C89 compiler I'm using doesn't support the unspecified array size within a struct - it gives an error. So I just made it use 1 and intended to reduce the size of the allocated space, but then forgot. It's not a huge problem that the buffer's a little larger, but if you want I can fix that up in the rest of the code?

Cyberspice reviewed (commented):
Good work. I assume the abstraction is because RISC OS treats block devices separately from file devices where as *NIX operating systems do not (which is why I just used FILE*). There was one thing that jumped out at me so I commented on it. Thanks for contributing.

gerph commented:
> Good work. I assume the abstraction is because RISC OS treats block devices separately from file devices where as _NIX operating systems do not (which is why I just used FILE_). There was one thing that jumped out at me so I commented on it. Thanks for contributing.

Yup, that's exactly right. I'm intending to take the non-command line bits and to build a very rudimentary system that uses the block access system in RISC OS. It's probably not something that most people would need, and you just wouldn't bother with this abstraction otherwise, but ... well, that's the way RISC OS is!

gerph committed:
Add renovate (mend.io) processing of this fork

gerph committed:
Add PR/issue templates.

Make it easier to work with by adding some templates for PR and issues.

renovate[bot] committed:
Update actions/checkout action to v6

renovate[bot] committed:
Update actions/download-artifact action to v7

gerph committed:
Merge pull request #1 from gerph/renovate/actions-checkout-6.x

Update actions/checkout action to v6

gerph committed:
Merge pull request #2 from gerph/renovate/actions-download-artifact-7.x

Update actions/download-artifact action to v7

renovate[bot] committed:
Update actions/upload-artifact action to v6

gerph committed:
Merge pull request #3 from gerph/renovate/actions-upload-artifact-6.x

Update actions/upload-artifact action to v6

renovate[bot] committed:
Update actions/download-artifact action to v8

renovate[bot] committed:
Update actions/upload-artifact action to v7

gerph committed:
Update issues to auto-assign

gerph committed:
Merge pull request #5 from gerph/renovate/actions-upload-artifact-7.x

Update actions/upload-artifact action to v7

gerph committed:
Merge pull request #4 from gerph/renovate/actions-download-artifact-8.x

Update actions/download-artifact action to v8

gerph committed:
Improve the build speed by using HostAMU.

Switch around the way that we build on the build service by using the
HostAMU command, rather than the RISC OS one.

renovate[bot] committed:
Update pozil/auto-assign-issue action to v3

gerph committed:
Merge pull request #6 from gerph/renovate/pozil-auto-assign-issue-3.x

Update pozil/auto-assign-issue action to v3

renovate[bot] committed:
Update pozil/auto-assign-issue action to v4

gerph committed:
Merge pull request #7 from gerph/renovate/pozil-auto-assign-issue-4.x

renovate[bot] committed:
Update actions/checkout action to v7

gerph committed:
Merge pull request #8 from gerph/renovate/actions-checkout-7.x

Update actions/checkout action to v7
